### ISSUE TITLE
IntegrityError: duplicate key value violates unique constraint "entity_locale_active"

### ISSUE THREAD

**Issue by bugzilla-to-github** [mozilla/pontoon#2284]
*This issue was created automatically by a [script](https://github.com/mathjazz/bugzilla2github/).*
## [Bug 1703666](https://bugzilla.mozilla.org/show_bug.cgi?id=1703666)
Bug Reporter: @mathjazz
CC:   @flodolo

An IntegrityError can appear during sync, which prevents the sync job to complete.

The only way to resolve the issue it to manually remove the conflicting active translation from Pontoon (by either deleting, rejecting or unapproving it).

```
Traceback (most recent call last):
File "/app/.heroku/python/lib/python3.8/site-packages/celery/app/trace.py", line 385, in trace_task
  R = retval = fun(*args, **kwargs)
File "/app/.heroku/python/lib/python3.8/site-packages/newrelic/hooks/application_celery.py", line 84, in wrapper
  return wrapped(*args, **kwargs)
File "/app/.heroku/python/lib/python3.8/site-packages/celery/app/trace.py", line 648, in __protected_call__
  return self.run(*args, **kwargs)
File "/app/pontoon/sync/core.py", line 68, in wrapped_func
  return func(self, *args, **kwargs)
File "/app/pontoon/sync/tasks.py", line 101, in sync_project
  sync_translations(
File "/app/pontoon/sync/tasks.py", line 280, in sync_translations
  changeset.execute()
File "/app/pontoon/sync/changeset.py", line 149, in execute
  Entity.objects.filter(
File "/app/pontoon/base/models.py", line 2451, in reset_active_translations
  translations.filter(Q(approved=True) | Q(fuzzy=True)).update(active=True)
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/models/query.py", line 784, in update
  rows = query.get_compiler(self.db).execute_sql(CURSOR)
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1522, in execute_sql
  cursor = super().execute_sql(result_type)
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1156, in execute_sql
  cursor.execute(sql, params)
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
  return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
  return executor(sql, params, many, context)
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
  return self.cursor.execute(sql, params)
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
  raise dj_exc_value.with_traceback(traceback) from exc_value
File "/app/.heroku/python/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
  return self.cursor.execute(sql, params)
File "/app/.heroku/python/lib/python3.8/site-packages/newrelic/hooks/database_psycopg2.py", line 34, in execute
  return super(CursorWrapper, self).execute(sql, parameters, *args,
File "/app/.heroku/python/lib/python3.8/site-packages/newrelic/hooks/database_dbapi2.py", line 24, in execute
  return self.__wrapped__.execute(sql, parameters,
django.db.utils.IntegrityError: duplicate key value violates unique constraint "entity_locale_active"
DETAIL:  Key (entity_id, locale_id, active)=(166302, 317, t) already exists.
```

**Comment by mathjazz** [comment]
The issue has become more common recently with the introduction of Pretranslation:
https://pontoon.mozilla.org/es-AR/mozillaorg/all-resources/?string=282842

**Comment by flodolo** [comment]
We have more errors

```
Nov 09 08:26:27 mozilla-pontoon app/web.1 django.db.utils.IntegrityError: duplicate key value violates unique constraint "entity_locale_active"
Nov 13 15:03:13 mozilla-pontoon app/web.3 django.db.utils.IntegrityError: duplicate key value violates unique constraint "entity_locale_active"
Nov 18 05:21:24 mozilla-pontoon app/worker.1 django.db.utils.IntegrityError: duplicate key value violates unique constraint "entity_locale_active"
(13 more at this point)
Nov 18 07:41:09 mozilla-pontoon app/worker.1 django.db.utils.IntegrityError: duplicate key value violates unique constraint "entity_locale_active"
```

**Comment by mathjazz** [comment]
> The issue has become more common recently with the introduction of Pretranslation: https://pontoon.mozilla.org/es-AR/mozillaorg/all-resources/?string=282842

In the last batch of new strings added to Mozilla.org, this type of errors only appeared in es-AR and only in pretranslations that contain Fluent message references and variables. It turns out that placeables in these pretranslations were not stored in a canonical form, i.e. `{example}` instead of `{ example }` (mind the whitespace). 

In the exported files in the repository, the strings appeared as they should, which resulted in sync trying to import pretranslations, becuse they weren't the same as in Pontoon DB. That also triggered this issue. In the UI (editor, string list) the strings were also rendered properly, which only made the problem more difficult to spot.

As the first step, we should make sure that re-serializing pretranslations always stores them in the canonical form. For the record, this is the code in question, which (when used from the command line) actually solved the immediate sync issue:
https://github.com/mozilla/pontoon/blob/73c05e4777bf2e78058743df86a02b0e30f54e57/pontoon/pretranslation/pretranslate.py#L59

Error could also be prevented by fixing #2120.

Finally, since the reason the problem appears only in es-AR could be that we have used TM entries with bad placeables formatting to train the MT engine, we should also look in that direction.

**Comment by mathjazz** [comment]
As the general fix for this very issue, we need to make sure translations actually get deactivated before new active translations are set:
https://github.com/mozilla/pontoon/blob/73c05e4777bf2e78058743df86a02b0e30f54e57/pontoon/base/models.py#L2613-L2619

The following resources might help understand and mitigate the problem:

https://docs.djangoproject.com/en/4.2/topics/db/transactions/#performing-actions-after-commit

My suspicion is this:
* we run the sync process within an [atomic transaction](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#django.db.transaction.atomic)
* that makes sure that all DB changes are rolled back if the sync fails at any point
* it also [prevents committing the DB query](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#django-s-default-transaction-behavior) to deactivate translations to the DB before new active translations are set
* that means that we can produce duplicate active translations when marking them as active 
* the solution might be to use [`on_commit()`](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#performing-actions-after-commit)

**Comment by flodolo** [comment]
> appears only in es-AR could be that we have used TM entries with bad placeables formatting to train the MT engine

Given the amount of linters we have in place, I don't think these would exist in the "active" translations (strings currently in repositories). But, from my experience, translation memory has a lot more that doesn't show up in Pontoon, so it's still a possibility.

**Comment by mathjazz** [comment]
This issue resurfaced in Hungarian Mozilla VPN Client after https://github.com/mozilla-l10n/mozilla-vpn-client-l10n/commit/2ab98c299c10eda1d32a041acfd28f76fc08b7dd landed.

We hit:

```
IntegrityError('duplicate key value violates unique constraint "entity_locale_active"
DETAIL:  Key (entity_id, locale_id, active)=(304357, 100, t) already exists.\n')

IntegrityError('duplicate key value violates unique constraint "entity_locale_active"
DETAIL:  Key (entity_id, locale_id, active)=(304326, 100, t) already exists.\n')

IntegrityError('duplicate key value violates unique constraint "entity_locale_active"
DETAIL:  Key (entity_id, locale_id, active)=(304322, 100, t) already exists.\n')
```

Approving failing Pretranslations didn't help. I had to delete all of them, sync again, and then run pretranslation.

**Comment by mathjazz** [comment]
The following order would be preferable:
https://github.com/mozilla/pontoon/issues/3904
https://github.com/mozilla/pontoon/issues/3924
https://github.com/mozilla/pontoon/issues/3802

**Comment by eemeli** [comment]
I don't understand what's going on with this issue, which probably should've been closed together with the sync refactor -- whatever might be happening now is quite likely to be due to different reasons than what was happening in 2021. It's also unclear to me if this is about an issue with sync, or an issue with pretranslation. It might be best to file a new issue about the specific current issue.

I'm also not sure what should be concluded from the duplicates listed in https://github.com/mozilla/pontoon/issues/2284#issuecomment-4188304804 above, as it doesn't filter out translations for which the entity strings are equal. Or why these counts are relevant for the IntegrityError in question.

**Comment by flodolo** [comment]
This is the last error we got (on April 3).
https://cloudlogging.app.goo.gl/XfGRkDBxiyig6gci6

I couldn't find an easy way to export the logs.

It is preceded by "Fetching pretranslations for locale es-AR started".

**Comment by eemeli** [comment]
I found one example in the logs:

```
2026-04-01 13:31:25.207Z  Fetching pretranslations for locale gl done
2026-04-01 13:31:25.207Z  Fetching pretranslations for locale de started
2026-04-01 13:31:50.531Z  Fetching pretranslations for locale de done
2026-04-01 13:31:50.531Z  Fetching pretranslations for locale pt-PT started
2026-04-01 13:31:55.923Z  Fetching pretranslations for project Firefox for Android started
2026-04-01 13:31:56.502Z  Fetching pretranslations for locale bg started
2026-04-01 13:31:59.713Z  Fetching pretranslations for locale bg done
2026-04-01 13:31:59.713Z  Fetching pretranslations for locale fr started
2026-04-01 13:32:02.806Z  Fetching pretranslations for locale fr done
2026-04-01 13:32:02.806Z  Fetching pretranslations for locale gl started
2026-04-01 13:32:05.974Z  Fetching pretranslations for locale gl done
2026-04-01 13:32:05.974Z  Fetching pretranslations for locale de started
2026-04-01 13:32:09.148Z  Fetching pretranslations for locale de done
2026-04-01 13:32:09.148Z  Fetching pretranslations for locale pt-PT started
2026-04-01 13:32:36.145Z  Fetching pretranslations for locale pt-PT done
2026-04-01 13:32:36.145Z  Fetching pretranslations for locale es-AR started
2026-04-01 13:32:39.020Z  duplicate key value violates unique constraint "entity_locale_active"
2026-04-01 13:32:39.020Z  DETAIL:  Key (entity_id, locale_id, active)=(331979, 186, t) already exists.
```

The specific error is due to a pre-existing `pt-PT` translation, which was created at 2026-04-01 13:32:35.821Z -- and note how there are two "Fetching pretranslations for locale pt-PT started" logs, from 13:31:50 and 13:32:09.

@mathjazz I think we're occasionally starting pretranslation twice for the same contents. Or otherwise ptetranslating the same contents twice.

**Comment by flodolo** [comment]
<img width="999" height="182" alt="Image" src="https://github.com/user-attachments/assets/d8375477-ba4e-4384-8c65-63341ddd8cfd" />


Possibly another argument in support of the theory that we have multiple operations running in parallel on the same content.
Before failing, the two syncs were both in progress.

**Comment by flodolo** [comment]
Hit this again for Firefox this morning. Sync was disabled, so there were no duplicated syncs happening.
